README.md
Description
===========

Installs the [Transmission BitTorrent Client](http://transmissionbt.org) and includes a `transmission_torrent_file` LWRP.

Requirements
============

Platform
--------

Known to work on Debian, Ubuntu, Red Hat and CentOS

Cookbooks
---------

* build-essential - for compiling the source.
* openssl - for generating a secure password.

Attributes
==========

* `node["transmission"]["peer_port"]` - The port Transmission listens on for peering, default `51413`.
* `node["transmission"]["rpc_bind_address"]` - Where to listen for RPC connections, default `0.0.0.0`.
* `node["transmission"]["rpc_port"]` - The port Transmission listens on for remote services, default `9091`.
* `node["transmission"]["rpc_username"]` - The username required to access remote services, default `transmission`.
* `node["transmission"]["rpc_password"]` - The ssha1 encrypted password, default generated by `secure_password`.
* `node["transmission"]["home"]` - Transmission's home directory, default `/var/lib/transmission-daemon`.
* `node["transmission"]["config_dir"]` - Transmission's config directory, default `/var/lib/transmission-daemon/info`.
* `node["transmission"]["download_dir"]` - Directory to move complete files to, default `/var/lib/transmission-daemon/downloads`.
* `node["transmission"]["incomplete_dir"]` - Directory to keep incomplete files in, default `/var/lib/transmission-daemon/incomplete`.
* `node["transmission"]["incomplete_dir_enabled"]` - Whether the incomplete file directory is used, default `false`.
* `node["transmission"]["speed_limit_down"]` - The KB/s limit for downloading files, default `100`.
* `node["transmission"]["speed_limit_down_enabled"]` - Whether the download limit is used, default `false`.
* `node["transmission"]["speed_limit_up"]` - The KB/s limit for uploading files, default `100`.
* `node["transmission"]["speed_limit_up_enabled"]` - Whether the upload limit is used, default `false`.

The file also contains the following attribute types:
* additional settings for transmisison.
* platform specific locations and settings.
* source installation settings.

Resource/Provider
=================

`transmission_torrent_file`
---------------------------

Download a file via the [BitTorrent protocol](http://en.wikipedia.org/wiki/BitTorrent).  The usage semantics are like that of the existing [file](http://wiki.opscode.com/display/chef/Resources#Resources-File) and [remote_file](http://wiki.opscode.com/display/chef/Resources#Resources-RemoteFile) resources.  This allows very fast downloads for files that are part of large BitTorrent swarms.  The Ubuntu 10.04 LTS ISO (around 700MB) downloads in about 50 seconds.

# Actions

- :create: Download a file via the BitTorrent protocol

# Attribute Parameters

- path: name attribute. the path to the file
- torrent: torrent file of the swarm to join.  can either be a url or local file path
- blocking: should the file be downloaded in a blocking way?  If `true` Chef will download the file in a single Chef run, if `false` Chef will check for a completed download during each Chef run until the download is complete. default is `true`.
- continue_seeding: should the file continue to be seeded to the swarm after download? default is `false`.
- owner: The owner for the file
- group: The group owner of the file (string or id)
- rpc_host: the address of the Transmission RPC host to connect to. default is `localhost`.
- rpc_port: the port of the Transmission RPC host to connect to. default is `9091`.
- rpc_username: the username of the Transmission RPC account. default is `transmission`.
- rpc_password: the password of the Transmission RPC account . should probably be `node['transmission']['rpc_password']` which by default is a secure password generated for this node.

# Example

    # download the lucid iso
    transmission_torrent_file "/home/ubuntu/ubuntu.iso"  do
      torrent "http://releases.ubuntu.com/lucid/ubuntu-10.04.1-server-i386.iso.torrent"
      owner 'ubuntu'
      group 'ubuntu'
      rpc_username node['transmission']['rpc_username']
      rpc_password node['transmission']['rpc_password']
      action :create
    end

    # do the same thing but continue seeding after download
    transmission_torrent_file "/home/ubuntu/ubuntu.iso"  do
      torrent "http://releases.ubuntu.com/lucid/ubuntu-10.04.1-server-i386.iso.torrent"
      owner 'ubuntu'
      group 'ubuntu'
      continue_seeding true
      rpc_username node['transmission']['rpc_username']
      rpc_password node['transmission']['rpc_password']
      action :create
    end

Usage
=====

default
-------

Include default recipe in a run list, to get some Transmission installed. Installs Transmission by package or source depending on the platform.

package
-------

Installs Transmission from packages.  This should only be loaded by the default recipe.

source
------

Installs Transmission from source.  This should only be loaded by the default recipe.

License and Author
==================

Author: Seth Chisamore (<schisamo@opscode.com>)

Copyright 2011, Opscode, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


CHANGELOG.md
## v1.2.0
* Exposed additional settings from Transmission
* We now create two copies of settings.json instead of trying to 
  symlink the file. Certain configurations on Ubuntu result in 
  apt creating a symlink as well -- in the oppostie direction -- 
  causing crashes when we run with chef-client.
* settings.json is now set to be owned by Transmission's group, 
  not root. This aligns with the default install from packages.
* settings.json is no longer world-readable by default; it 
  contains the RPC login, which needs to be protected.
* Added node['transmission']['settings_mode'] which controls the
  permissions on settings.json; the default is 660, allowing the
  settings to be changed by transmission from the web interface.
  If this is undesirable, the mode should be set to 640, 
  preventing changes to the configuration except by root/chef.



## v1.0.2:

* [COOK-729] - `transmission_torrent_file` doesn't work for more than
  a single torrent
* [COOK-732] - link to file in swarm not created if torrent already
  completely downloaded


attributes/default.rb
#
# Author:: Seth Chisamore (<schisamo@opscode.com>)
# Cookbook Name:: transmission
# Attribute:: default
#
# Copyright 2011, Opscode, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

::Chef::Node.send(:include, Opscode::OpenSSL::Password)

case node['platform']
when "ubuntu","debian"
  default['transmission']['install_method'] = 'package'
  default['transmission']['user']           = 'debian-transmission'
  default['transmission']['group']          = 'debian-transmission'
else
  default['transmission']['install_method'] = 'source'
  default['transmission']['user']           = 'transmission'
  default['transmission']['group']          = 'transmission'
end

default['transmission']['settings_mode']    = '660' #Controls the mode of settings.json
                                                    # '660' allows the transmission daemon to update the settings, e.g. from the web interface
                                                    # '640' only allows changes by root/chef
                                                    # This file should never be world readable; it contains the RPC password.

default['transmission']['url']              = 'http://download.transmissionbt.com/files'
default['transmission']['version']          = '2.03'
default['transmission']['checksum']         = '06802c6f4ba517341eb287b273145ccd5e7b0fba2a270da82f0eb0a683cf4046'

default['transmission']['bind_ipv4']               = "0.0.0.0"
default['transmission']['bind_ipv6']               = "::"
default['transmission']['encryption']              = 1 #(0=Prefer unencrypted, 1=Prefer encrypted, 2=Require encrypted)
default['transmission']['dht_enabled']             = 'true'
default['transmission']['blocklist_enabled']       = 'false' #This is mostly useless these days, but here if you want it
default['transmission']['blocklist_url']           = "http://www.example.com/blocklist" #You need to change this if enabling blocklist

default['transmission']['peer_port']               = 51413
default['transmission']['peer_port_random']        = 'false'
default['transmission']['peer_port_random_min']    = 49152
default['transmission']['peer_port_random_max']    = 65535
default['transmission']['peer_socket_tos']         = "default"
default['transmission']['peer_port_forward']       = 'false'

default['transmission']['peer_exchange']    = 'true'

default['transmission']['rpc_enabled'] = 'true'
default['transmission']['rpc_whitelist_enabled'] = 'true'
default['transmission']['rpc_whitelist'] = '127.0.0.1'
default['transmission']['rpc_bind_address'] = '0.0.0.0'
default['transmission']['rpc_auth_required'] = 'true'
default['transmission']['rpc_username']     = 'transmission'
set_unless['transmission']['rpc_password']  = secure_password
default['transmission']['rpc_port']         = 9091

default['transmission']['umask']            = 18
default['transmission']['home']             = '/var/lib/transmission-daemon'
default['transmission']['config_dir']       = '/var/lib/transmission-daemon/info'
default['transmission']['download_dir']     = '/var/lib/transmission-daemon/downloads'
default['transmission']['incomplete_dir']   = '/var/lib/transmission-daemon/incomplete'
default['transmission']['incomplete_dir_enabled'] = 'false'

#Baseline speed limits, these are the absolute highest that should be permitted
default['transmission']['speed_limit_down']         = 1000 #KB/s
default['transmission']['speed_limit_down_enabled'] = 'false'
default['transmission']['speed_limit_up']           = 1000 #KB/s
default['transmission']['speed_limit_up_enabled']   = 'false'
#These are 'alternate' a.k.a. 'daytime' speed limits to help you avoid pissing off your local Network Admin :)

default['transmission']['alt_speed_limit_enabled']  = 'false' #this controls both alt-speed-time-enabled and alt-speed-enabled 
                                                              # since from a conf standpoint they should both start on or off, 
                                                              # then alt-speed-enabled should vary according to time.
default['transmission']['alt_speed_limit_down']     = 1000 #KB/s
default['transmission']['alt_speed_limit_up']       = 100 #KB/s
default['transmission']['alt_speed_limit_begin']    = 540 #(default = 540, in minutes from midnight, 09:00)
default['transmission']['alt_speed_limit_end']      = 1020 #(default = 1020, in minutes from midnight, 17:00)
default['transmission']['alt_speed_limit_day']      = 127 #(default = 127, all days)
                                                          # Start with 0, then for each day you want the scheduler enabled, add:
                                                          #   Sunday: 1 (binary: 0000001)
                                                          #   Monday: 2 (binary: 0000010)
                                                          #   Tuesday: 4 (binary: 0000100)
                                                          #   Wednesday: 8 (binary: 0001000)
                                                          #   Thursday: 16 (binary: 0010000)
                                                          #   Friday: 32 (binary: 0100000)
                                                          #   Saturday: 64 (binary: 1000000)
                                                          # Examples:
                                                          #   Weekdays: 62 (binary: 0111110)
                                                          #   Weekends: 65 (binary: 1000001)
                                                          #   All Days: 127 (binary: 1111111)


default['transmission']['peer_limit']               = 200
default['transmission']['peer_limit_global']        = 240
default['transmission']['peer_limit_torrent']       = 60
